### PR TITLE
[NativeAOT-LLVM] Add support CT_INDIRECT calls

### LINKED ISSUE #1862: [NativeAOT-LLVM] [Question] consuming return values from CT_INDIRECT
Looking at this IR:
```
                                                  /--*  t84    int    arg-1
                                                  +--*  t32    int    calli tgt
N006 ( 26, 11) [000029] --CXG-------        t29 = *  CALL ind  int
                                                  /--*  t29    int
N008 ( 30, 14) [000027] DA-XG-------              *  STORE_LCL_VAR int    V05 tmp3         d:4```
```
`t29` is stored in the sdsu map as an `i8`, and then fails to be consumed to an `int` for the `STORE_LCL_VAR`.

@SingleAccretion, previously you wrote

> To maintain the invariant under 1 (only exact types stored in locations), we are to pass the type of the local to consumeValue (and ignore the node type).

Is this the erroneous logic in `consumeValue` at

https://github.com/dotnet/runtimelab/blob/258c8cee8cec40849127995fee955c117ced3d42/src/coreclr/jit/llvm.cpp#L923

Should that be using `targetLlvmType` which comes from the type of the local?

### PR DESCRIPTION
**Pull request by yowl**
This PR adds support for CT_INDIRECT calls.  It promotes `GenTree`'s `callSig` to a non DEBUG only field for `TARGET_WASM`   Doesn't add a huge amount of methods to the clrjit coverage, about a 1% increase.

cc @SingleAccretion

Closes #1862  #1860

### CODE REVIEW

File: src/coreclr/jit/llvm.cpp [hunk 1]
```diff
@@ -919,8 +919,10 @@ Value* Llvm::consumeValue(GenTree* node, Type* targetLlvmType)
                nodeValue->getType()->getPrimitiveSizeInBits() <= 32 && targetLlvmType->getPrimitiveSizeInBits() <= 32);
         if (nodeValue->getType()->getPrimitiveSizeInBits() < targetLlvmType->getPrimitiveSizeInBits())
         {
+            var_types trueNodeType = node->OperIs(GT_CALL) ? static_cast<var_types>(node->AsCall()->gtReturnType) : node->TypeGet();
```
Comment (SingleAccretion) [inline src/coreclr/jit/llvm.cpp]: I think we can convert the whole chain of `if ... else`s in the method to this now:
```cs
var_types trueNodeType = TYP_UNDEF;
switch (node->OperGet())
    case GT_CALL: { };
    case GT_LCL_VAR: { };
    case GT_[RELOPS]: { };
    default: trueNodeType = node->TypeGet();

assert(varTypeIsSmall(trueNodeType));

finalValue = varTypeIsSigned(trueNodeType) ? _builder.CreateSExt(nodeValue, targetLlvmType)
                                           : _builder.CreateZExt(nodeValue, targetLlvmType);
```
(Also, right now the case under `// Upcast.` is using `varTypeIsSigned(node)`, it should use `varTypeIsSigned(trueNodeType)`)

File: src/coreclr/jit/llvm.cpp [hunk 2]
```diff
@@ -2399,12 +2400,13 @@ void Llvm::lowerToShadowStack()
...
                 }
 
-                CORINFO_SIG_INFO calleeSigInfo;
-                _compiler->eeGetMethodSig(callNode->gtCallMethHnd, &calleeSigInfo);
-
-                failUnsupportedCalls(callNode, &calleeSigInfo);
+                if (callNode->gtCallType == CT_INDIRECT)
+                {
+                    unsigned i = 1;
+                }
```
Comment (SingleAccretion) [inline src/coreclr/jit/llvm.cpp]:
```suggestion
```

File: src/coreclr/jit/llvm.cpp (hunk 1 above)
Comment (yowl) [reply src/coreclr/jit/llvm.cpp]: I made a change here .  For Relops I did that in the default block, not seeing a way to make it a `case`.

I also added a couple of more sets for the `callSig`.  The clone was the one breaking the release tests I think.  The `CORINFO_VIRTUALCALL_LDVIRTFTN` case also looked like it needs it.

File: src/coreclr/jit/llvm.cpp (hunk 2 above)
Comment (yowl) [reply src/coreclr/jit/llvm.cpp]: Thanks

File: src/coreclr/jit/gentree.cpp [hunk 3]
```diff
@@ -8799,7 +8799,11 @@ GenTreeCall* Compiler::gtCloneExprCallHelper(GenTreeCall* tree,
     // we only really need one physical copy of it. Therefore a shallow pointer copy will suffice.
     // (Note that this still holds even if the tree we are cloning was created by an inlinee compiler,
     // because the inlinee still uses the inliner's memory allocator anyway.)
+#if TARGET_WASM
+    copy->callSig = tree->callSig;
+#else
     INDEBUG(copy->callSig = tree->callSig;)
+#endif // TARGET_WASM
```
Comment (SingleAccretion) [inline src/coreclr/jit/gentree.cpp]:
```suggestion
#if defined(DEBUG) || defined(TARGET_WASM)
    copy->callSig = tree->callSig;
#endif // DEBUG || TARGET_WASM
```

File: src/coreclr/jit/llvm.cpp [hunk 4]
```diff
@@ -919,28 +919,40 @@ Value* Llvm::consumeValue(GenTree* node, Type* targetLlvmType)
...
+                    {
+                        // This is the special case for relops. Ordinary codegen "just knows" they need zero-extension.
+                        assert(nodeValue->getType() == Type::getInt1Ty(_llvmContext));
+                        trueNodeType = TYP_BOOL;
+                    }
+                    else
+                    {
+                        trueNodeType = node->TypeGet();
+                    }
+                }
```
Comment (SingleAccretion) [inline src/coreclr/jit/llvm.cpp]:
```suggestion
                case GT_CALL:
                    trueNodeType = static_cast<var_types>(node->AsCall()->gtReturnType);
                    break;

                case GT_LCL_VAR:
                    trueNodeType = _compiler->lvaGetDesc(node->AsLclVarCommon())->TypeGet();
                    break;

                case GT_EQ:
                case GT_NE:
                case GT_LT:
                case GT_LE:
                case GT_GE:
                case GT_GT:
                    // This is the special case for relops. Ordinary codegen "just knows" they need zero-extension.
                    assert(nodeValue->getType() == Type::getInt1Ty(_llvmContext));
                    trueNodeType = TYP_UBYTE;
                    break;

                default:
                    trueNodeType = node->TypeGet();
                    break;
```
(The relops can be just listed explicitly, we'll not be adding new ones any time soon)

File: src/coreclr/jit/importer.cpp [hunk 5]
```diff
@@ -8644,6 +8644,11 @@ var_types Compiler::impImportCall(OPCODE                  opcode,
                 // Sine we are jumping over some code, check that its OK to skip that code
                 assert((sig->callConv & CORINFO_CALLCONV_MASK) != CORINFO_CALLCONV_VARARG &&
                        (sig->callConv & CORINFO_CALLCONV_MASK) != CORINFO_CALLCONV_NATIVEVARARG);
+
+#if TARGET_WASM // TODO-LLVM
+                call->AsCall()->callSig  = new (this, CMK_Generic) CORINFO_SIG_INFO;
+                *call->AsCall()->callSig = *sig;
```
Comment (SingleAccretion) [inline src/coreclr/jit/importer.cpp]:
```suggestion
                call->AsCall()->callSig  = new (this, CMK_Generic) CORINFO_SIG_INFO(*sig);
```
(It *should* work, though I did not test it)

File: src/coreclr/jit/importer.cpp [hunk 6]
```diff
@@ -8644,6 +8644,11 @@ var_types Compiler::impImportCall(OPCODE                  opcode,
                 // Sine we are jumping over some code, check that its OK to skip that code
                 assert((sig->callConv & CORINFO_CALLCONV_MASK) != CORINFO_CALLCONV_VARARG &&
                        (sig->callConv & CORINFO_CALLCONV_MASK) != CORINFO_CALLCONV_NATIVEVARARG);
+
+#if TARGET_WASM // TODO-LLVM
```
Comment (SingleAccretion) [inline src/coreclr/jit/importer.cpp]: Why the `TODO`?

File: src/coreclr/jit/importer.cpp [hunk 7]
```diff
@@ -8737,6 +8742,12 @@ var_types Compiler::impImportCall(OPCODE                  opcode,
     assert(((mflags & CORINFO_FLG_STATIC) != 0) == ((sig->callConv & CORINFO_CALLCONV_HASTHIS) == 0));
     assert(call != nullptr);
 
+#if TARGET_WASM
+    assert(call->AsCall()->callSig == nullptr);
+    call->AsCall()->callSig  = new (this, CMK_Generic) CORINFO_SIG_INFO;
+    *call->AsCall()->callSig = *sig;
```
Comment (SingleAccretion) [inline src/coreclr/jit/importer.cpp]:
```suggestion
    call->AsCall()->callSig  = new (this, CMK_Generic) CORINFO_SIG_INFO(*sig);
```

**Review by SingleAccretion — APPROVED** [review]
Just some code simplification suggestions, otherwise LGTM.

File: src/coreclr/jit/importer.cpp (hunk 6 above)
Comment (yowl) [reply src/coreclr/jit/importer.cpp]: Dont remember, I'll remove it.

File: src/coreclr/jit/importer.cpp (hunk 5 above)
Comment (yowl) [reply src/coreclr/jit/importer.cpp]: Thanks, looks fine here.

File: src/coreclr/jit/llvm.cpp (hunk 4 above)
Comment (yowl) [reply src/coreclr/jit/llvm.cpp]: Thanks

**Comment by yowl** [comment]
Oops, removed the #ifs by mistake

**Review by jkotas — APPROVED** [review]
Thank you!

**Pull request merged**